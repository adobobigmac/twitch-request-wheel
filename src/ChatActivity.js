const fetch = require('node-fetch');

export const ActivityStatus = {
  ACTIVE: 1,
  IDLE: 2,
  DISCONNECTED: 3
};

const MAX_IDLE_TIME_MINUTES = 10;

export default class ChatActivity {
  constructor(channel) {
    this.channel = channel;
    this.lastMessageTimes = {};
    this.getStatus = this.getStatus.bind(this);
  }

  updateLastMessageTime = (user) => {
    this.lastMessageTimes = {
      ...this.lastMessageTimes,
      [user]: Date.now()
    };
  }

  minsSinceLastChatMessage = (user) => {
    return Math.floor((Date.now()-this.lastMessageTimes[user])/60000);
  }

  async getStatus(user) {
    // broadcaster always counts as active
    if(user === this.channel) {
      return ActivityStatus.ACTIVE;
    }

    // sent a chat message in the last X mins?
    if(this.lastMessageTimes[user] && this.minsSinceLastChatMessage(user) < MAX_IDLE_TIME_MINUTES) {
      return ActivityStatus.ACTIVE;
    }

    // check the twitch api to see if they have an active connection to chat.
    // (thank the lord almighty for free open proxy sites. CORB is annoying.)
    // note that results are fairly heavily cached, and the API may break
    // eventually (is undocumented), but this is what twitch themselves
    // uses to display the list of people connected to chat - best we got.
    return fetch(`https://thingproxy.freeboard.io/fetch/https://tmi.twitch.tv/group/user/${this.channel}/chatters`)
      .then(r => r.json())
      .then(res => {
        console.log(res);
        if(!res.chatters) {
          return ActivityStatus.DISCONNECTED; // in case the response doesn't parse goodly
}
        if(res.chatters.vips.includes(user)
           || res.chatters.moderators.includes(user)
           || res.chatters.viewers.includes(user)
           || res.chatters.staff.includes(user)
           || res.chatters.admins.includes(user)
           || res.chatters.global_mods.includes(user)
        ) {
          console.log("IDLEEE")
          return ActivityStatus.IDLE;
        }
        return ActivityStatus.DISCONNECTED;
      })
      .catch((e) => {
        console.log(e);
      })
  }
}
