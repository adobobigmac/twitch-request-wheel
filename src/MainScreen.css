#main-screen {
    display: flex;
}
#main-screen.game-select .left-column {
    width: 45vw;
}
#main-screen.player-select .left-column {
    width: 90vw;
}

#main-screen .left-column h2 {
    margin-bottom: 0;
}
#main-screen .left-column h4 {
    color: yellow;
    font-size: 20px;
    font-weight:  400;
    margin-bottom: 12px;
    margin-top: 6px;
}
#main-screen.player-select .subheading-game,
#main-screen.game-select .subheading-player {
    display: none;
}

.left-column .left-column-body {
    display: flex;
    align-items: flex-start;
    height: 100%;
}
.left-column .left-column-inner-body {
    flex-grow: 2;
    margin-left: 15px;
}

#main-screen .right-column {
    /* width: 50vw; */
    text-transform: capitalize;
}

.wheel-wrapper {
    font-size: 16px;
    overflow: hidden;
    width: 600px;
}
.wheel-wrapper canvas {
    cursor: pointer;
}

#main-screen .logout {
    position: absolute;
    top:  0;
    right:  0;
    background-color: darkcyan;
    border-radius: 5px;
    margin-top:  0;
    padding-bottom: 5px;
    padding-top: 5px;
    color: #fff;
}



#main-screen .overlay {
    position: fixed;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    background-color: rgba(0,0,0,0.5);
}
#main-screen .confetti-wrapper {
    position: absolute;
    right: 50%;
    left: 50%;
    bottom: 50;
}
#main-screen .fade-in-out {
    opacity: 0;
    animation: 3.5s linear 0s main-modal-fade;
    animation-fill-mode: forwards;
}
#main-screen .modal {
    position: fixed;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
    background: rgba(0,0,0,0.5);
    padding: 2vh 4vw;
    color: #fff;
}
@keyframes main-modal-fade {
    0%,100% { opacity: 0 }
    5%,95% { opacity: 1 }
}
