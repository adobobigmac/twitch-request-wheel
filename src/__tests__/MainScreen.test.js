import {ActivityStatus} from '../ChatActivity';
import {createRenderer} from 'react-test-renderer/shallow';
import MainScreen from '../MainScreen';
import React from 'react';

jest.mock('react-wheel-of-prizes', () => {
    return {
        __esModule: true,
        default: () => {
            return <></>;
        },
        WheelComponent: 'WheelComponent'
    };
});

describe('MainScreen', () => {
    const props = {
        access_token: 'foo',
        channel: 'sirgoosewell',
        modList: [],
        onLogout: jest.fn()
    };

    describe('onWheelSpun', () => {
        const state = {
            messages: {
                'Trivia Murder Party 2': {
                    name: 'Trivia Murder Party 2',
                    longName: 'Trivia Murder Party 2',
                    username: 'aurora88877',
                    locked: false
                },
                'Quiplash 3': {
                    name: 'Quiplash 3',
                    longName: 'Quiplash 3',
                    username: 'johnell75',
                    locked: false
                },
                'Survive The Internet (Party Pack 4)': {
                    name: 'Survive The Internet',
                    longName: 'Survive The Internet (Party Pack 4)',
                    partyPack: 'Party Pack 4',
                    'Min players': 3,
                    'Max players': 8,
                    Variants: ['survive', 'survive the internet', 'sti'],
                    username: 'aurora88877',
                    time: 1627509347466,
                    locked: true,
                    chosen: false
                }
            },
            colors: ['#99b0fc', '#b4a7f2', '#beb4f7', '#cdc2f9', '#aec2f9', '#a4d6f9'],
            counter: 1,
            history: [],
            nextGameIdx: 0,
            showPlayerSelect: false
        };

        test('should respond to ActivityStatus when game found in state', async () => {
            jest.useFakeTimers();

            const component = new MainScreen(props);
            component.state = state;
            component.messageHandler = {
                sendMessage: jest.fn().mockImplementation((msg) => {
                    return Promise.resolve(msg);
                })
            };

            jest.spyOn(component.messageHandler, 'sendMessage').mockImplementation(() => {})
            jest.spyOn(component.chatActivity, 'getStatusPromise')
                .mockResolvedValueOnce(ActivityStatus.ACTIVE)
                .mockResolvedValueOnce(ActivityStatus.IDLE)
                .mockResolvedValueOnce(ActivityStatus.DISCONNECTED);
            jest.spyOn(component, 'addGameToQueue').mockImplementation(() => {});
            // jest.spyOn(component, 'removeGame').mockImplementation(() => {});
            jest.spyOn(component, 'setState').mockImplementation(() => {});

            await component.onWheelSpun('Trivia Murder Party 2');
            await component.onWheelSpun('Quiplash 3');
            await component.onWheelSpun('Survive The Internet (Party Pack 4)');
            await component.onWheelSpun('Quiplash 2');

            jest.advanceTimersByTime(5000);
            jest.clearAllTimers();

            expect(setTimeout).toHaveBeenCalledTimes(2);
            // expect(component.removeGame).toHaveBeenCalledTimes(2);
            expect(component.setState).toHaveBeenCalledTimes(2);
            expect(component.setState.mock.calls[0][0]()).toEqual({
                messages : {
                    'Quiplash 3': {
                        name: 'Quiplash 3',
                        longName: 'Quiplash 3',
                        username: 'johnell75',
                        locked: false
                    },
                    'Survive The Internet (Party Pack 4)': {
                        name: 'Survive The Internet',
                        longName: 'Survive The Internet (Party Pack 4)',
                        partyPack: 'Party Pack 4',
                        'Min players': 3,
                        'Max players': 8,
                        Variants: ['survive', 'survive the internet', 'sti'],
                        username: 'aurora88877',
                        time: 1627509347466,
                        locked: true,
                        chosen: false
                    }
                },
                counter: 2
            });

            expect(component.chatActivity.getStatusPromise).toHaveBeenCalledTimes(3);

            expect(component.messageHandler.sendMessage).toHaveBeenCalledTimes(3);
            expect(component.messageHandler.sendMessage).toHaveBeenNthCalledWith(1,
                expect.stringContaining('/me @aurora88877, good news')
            );
            expect(component.messageHandler.sendMessage).toHaveBeenNthCalledWith(2,
                expect.stringContaining('/me @johnell75, good news')
            );
            expect(component.messageHandler.sendMessage).toHaveBeenNthCalledWith(3,
                expect.stringContaining('/me Survive The Internet just won the spin')
            );
        });
    });
    describe('setMessageHandlerRef', () => {
        test('should set messageHandler instance', () => {
            const component = new MainScreen(props);
            let ref = React.createRef();
            component.setMessageHandlerRef(ref);
            expect(component.messageHandler).toBe(ref);
        });
    });
    describe('setPlayerSelectRef', () => {
        test('should set playerSelector instance', () => {
            const component = new MainScreen(props);
            let ref = React.createRef();
            component.setPlayerSelectRef(ref);
            expect(component.playerSelector).toBe(ref);
        });
    });
    describe('render', () => {
        test('renders page', () => {
            const shallowRenderer = createRenderer();
            shallowRenderer.render(<MainScreen/>);
            let component = shallowRenderer.getRenderOutput();
            expect(component.type).toBe('div');
            shallowRenderer.unmount();
        });
        test('should render the game request screen', () => {
            let state = {
            	messages: {
            		'Survive The Internet (Party Pack 4)': {
            			name: 'Survive The Internet',
            			longName: 'Survive The Internet (Party Pack 4)',
            			partyPack: 'Party Pack 4',
            			'Min players': 3,
            			'Max players': 8,
            			Variants: ['survive', 'survive the internet', 'sti'],
            			username: 'dcpesses',
            			time: 1627509347466,
            			locked: false,
            			chosen: false
            		}
            	},
            	colors: ['#99b0fc', '#b4a7f2', '#beb4f7', '#cdc2f9', '#aec2f9', '#a4d6f9'],
            	counter: 1,
            	history: [],
            	nextGameIdx: 0,
            	showPlayerSelect: false
            };
            const shallowRenderer = createRenderer();
            shallowRenderer.render(<MainScreen {...props}/>);
            let instance = shallowRenderer.getMountedInstance();
            instance.setState(state);
            let component = shallowRenderer.getRenderOutput();
            expect(component).toMatchSnapshot();
            // console.log(component.props.children.filter(p => p.type === 'button'));
            let logoutBtn = component.props.children.find(p => p.type === 'button');
            logoutBtn.props.onClick();
            // userEvent.click(logoutBtn);
            expect(props.onLogout).toHaveBeenCalledTimes(1);
            shallowRenderer.unmount();
        });
        test('should render the player selection screen', () => {
            let state = {
                messages: {
                    'Survive The Internet (Party Pack 4)': {
                        name: 'Survive The Internet',
                        longName: 'Survive The Internet (Party Pack 4)',
                        partyPack: 'Party Pack 4',
                        'Min players': 3,
                        'Max players': 8,
                        Variants: ['survive', 'survive the internet', 'sti'],
                        username: 'dcpesses',
                        time: 1627509347466,
                        locked: false,
                        chosen: false
                    }
                },
                colors: ['#99b0fc', '#b4a7f2', '#beb4f7', '#cdc2f9', '#aec2f9', '#a4d6f9'],
                counter: 1,
                history: [],
                nextGameIdx: 0,
                showPlayerSelect: true
            };
            const shallowRenderer = createRenderer();
            shallowRenderer.render(<MainScreen {...props}/>);
            let instance = shallowRenderer.getMountedInstance();
            instance.setState(state);
            expect(shallowRenderer.getRenderOutput()).toMatchSnapshot();
            expect(instance.playerSelector).toBeDefined();
            shallowRenderer.unmount();
        });
    });
});
