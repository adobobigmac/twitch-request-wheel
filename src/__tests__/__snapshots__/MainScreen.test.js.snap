// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainScreen render renders page 1`] = `
<div>
  <div
    class="game-select"
    id="main-screen"
  >
    <div
      class="left-column"
    >
      <h2>
        Game Requests
      </h2>
      <h4>
        <span
          class="subheading-game"
        >
          Type e.g. 
          <b>
            "!request Blather Round"
          </b>
           in 
          sirgoosewell
          's chat to add
        </span>
      </h4>
      <div
        class="left-column-body"
      >
        <div
          class="game"
          id="sidebar"
        >
          <div
            class="sidebar-panel"
          >
            <div
              class="p sidebar-panel-up-next"
            >
               Up Next:
              <p
                class="next-game-name"
              >
                not yet decided
              </p>
              <p
                class="next-game-party-pack"
              >
                <i />
              </p>
              <button
                class="open-seat-requests"
                disabled=""
              >
                OPEN SEAT REQUESTS
              </button>
            </div>
          </div>
          <div
            class="nav-wrapper"
          >
            <button
              class="move-next-game-back"
            >
               ⇦ 
            </button>
            <button
              class="move-next-game-fwd"
            >
               ⇨ 
            </button>
          </div>
          <div
            class="sidebar-panel"
          >
            <p
              class="sidebar-panel-history"
            >
               History 
            </p>
            <p
              class="sidebar-panel-history-list"
            >
              <li>
                No games yet
              </li>
            </p>
          </div>
        </div>
        <div
          class="left-column-inner-body"
        />
      </div>
    </div>
    <div
      class="right-column"
      width="50px"
    >
      <div
        class="wheel-wrapper"
      />
    </div>
  </div>
</div>
`;

exports[`MainScreen render should render the game request screen 1`] = `
<div
  className="game-select"
  id="main-screen"
>
  <MessageHandler
    access_token="foo"
    addGameRequest={[Function]}
    caniplayHandler={[Function]}
    changeNextGameIdx={[Function]}
    channel="sirgoosewell"
    clearQueueHandler={[Function]}
    closeQueueHandler={[Function]}
    messages={
      Object {
        "Survive The Internet (Party Pack 4)": Object {
          "Max players": 8,
          "Min players": 3,
          "Variants": Array [
            "survive",
            "survive the internet",
            "sti",
          ],
          "chosen": false,
          "locked": false,
          "longName": "Survive The Internet (Party Pack 4)",
          "name": "Survive The Internet",
          "partyPack": "Party Pack 4",
          "time": 1627509347466,
          "username": "dcpesses",
        },
      }
    }
    modList={Array []}
    onDelete={[Function]}
    onMessage={[Function]}
    openQueueHandler={[Function]}
    playerExitHandler={[Function]}
    setNextGame={[Function]}
    startGame={[Function]}
    upcomingGames={Array []}
  />
  <div
    className="left-column"
  >
    <h2>
      Game Requests
    </h2>
    <h4>
      <span
        className="subheading-game"
      >
        Type e.g. 
        <b>
          "!request Blather Round"
        </b>
         in 
        sirgoosewell
        's chat to add
      </span>
    </h4>
    <div
      className="left-column-body"
    >
      <Sidebar
        changeNextGameIdx={[Function]}
        history={Array []}
        moveNextGameBack={[Function]}
        moveNextGameFwd={[Function]}
        nextGameIdx={0}
        requestMode="game"
        togglePlayerSelect={[Function]}
      />
      <div
        className="left-column-inner-body"
      >
        <GameRequest
          gameName="Survive The Internet (Party Pack 4)"
          getActivity={[Function]}
          metadata={
            Object {
              "Max players": 8,
              "Min players": 3,
              "Variants": Array [
                "survive",
                "survive the internet",
                "sti",
              ],
              "chosen": false,
              "locked": false,
              "longName": "Survive The Internet (Party Pack 4)",
              "name": "Survive The Internet",
              "partyPack": "Party Pack 4",
              "time": 1627509347466,
              "username": "dcpesses",
            }
          }
          onDelete={[Function]}
          toggleLock={[Function]}
        />
      </div>
    </div>
  </div>
  <div
    className="right-column"
    width="50px"
  >
    <div
      className="wheel-wrapper"
    >
      <default
        contrastColor="black"
        downDuration={1000}
        fontFamily="Arial"
        isOnlyOnce={false}
        onFinished={[Function]}
        primaryColor="white"
        segColors={
          Array [
            "#99b0fc",
            "#b4a7f2",
            "#beb4f7",
            "#cdc2f9",
            "#aec2f9",
            "#a4d6f9",
          ]
        }
        segments={
          Array [
            "Survive The Internet (Party Pack 4)",
          ]
        }
        size={250}
        upDuration={100}
      />
    </div>
  </div>
  <button
    className="logout"
    onClick={[MockFunction]}
  >
    Logout ➧
  </button>
</div>
`;

exports[`MainScreen render should render the player selection screen 1`] = `
<div
  className="player-select"
  id="main-screen"
>
  <MessageHandler
    access_token="foo"
    addGameRequest={[Function]}
    caniplayHandler={[Function]}
    changeNextGameIdx={[Function]}
    channel="sirgoosewell"
    clearQueueHandler={[Function]}
    closeQueueHandler={[Function]}
    messages={
      Object {
        "Survive The Internet (Party Pack 4)": Object {
          "Max players": 8,
          "Min players": 3,
          "Variants": Array [
            "survive",
            "survive the internet",
            "sti",
          ],
          "chosen": false,
          "locked": false,
          "longName": "Survive The Internet (Party Pack 4)",
          "name": "Survive The Internet",
          "partyPack": "Party Pack 4",
          "time": 1627509347466,
          "username": "dcpesses",
        },
      }
    }
    modList={Array []}
    onDelete={[Function]}
    onMessage={[Function]}
    openQueueHandler={[Function]}
    playerExitHandler={[Function]}
    setNextGame={[Function]}
    startGame={[Function]}
    upcomingGames={Array []}
  />
  <div
    className="left-column"
  >
    <h2>
      Seat Requests
    </h2>
    <h4>
      <span
        className="subheading-player"
      >
        Type 
        <b>
          !new
        </b>
         in 
        sirgoosewell
        's chat if you want to join the next game
      </span>
    </h4>
    <div
      className="left-column-body"
    >
      <Sidebar
        changeNextGameIdx={[Function]}
        history={Array []}
        moveNextGameBack={[Function]}
        moveNextGameFwd={[Function]}
        nextGameIdx={0}
        requestMode="seat"
        togglePlayerSelect={[Function]}
      />
      <div
        className="left-column-inner-body"
      >
        <PlayerSelect
          startGame={[Function]}
        />
      </div>
    </div>
  </div>
  <button
    className="logout"
    onClick={[MockFunction]}
  >
    Logout ➧
  </button>
</div>
`;

exports[`MainScreen renders page 1`] = `
<div>
  <div
    class="game-select"
    id="main-screen"
  >
    <div
      class="left-column"
    >
      <h2>
        Game Requests
      </h2>
      <h4>
        <span
          class="subheading-game"
        >
          Type e.g. 
          <b>
            "!request Blather Round"
          </b>
           in 
          sirgoosewell
          's chat to add
        </span>
      </h4>
      <div
        class="left-column-body"
      >
        <div
          class="game"
          id="sidebar"
        >
          <div
            class="sidebar-panel"
          >
            <div
              class="p sidebar-panel-up-next"
            >
               Up Next:
              <p
                class="next-game-name"
              >
                not yet decided
              </p>
              <p
                class="next-game-party-pack"
              >
                <i />
              </p>
              <button
                class="open-seat-requests"
                disabled=""
              >
                OPEN SEAT REQUESTS
              </button>
            </div>
          </div>
          <div
            class="nav-wrapper"
          >
            <button
              class="move-next-game-back"
            >
               ⇦ 
            </button>
            <button
              class="move-next-game-fwd"
            >
               ⇨ 
            </button>
          </div>
          <div
            class="sidebar-panel"
          >
            <p
              class="sidebar-panel-history"
            >
               History 
            </p>
            <p
              class="sidebar-panel-history-list"
            >
              <li>
                No games yet
              </li>
            </p>
          </div>
        </div>
        <div
          class="left-column-inner-body"
        />
      </div>
    </div>
    <div
      class="right-column"
      width="50px"
    >
      <div
        class="wheel-wrapper"
      />
    </div>
  </div>
</div>
`;

exports[`MainScreen should render the game request screen 1`] = `
<div
  className="game-select"
  id="main-screen"
>
  <MessageHandler
    access_token="foo"
    addGameRequest={[Function]}
    caniplayHandler={[Function]}
    changeNextGameIdx={[Function]}
    channel="sirgoosewell"
    clearQueueHandler={[Function]}
    closeQueueHandler={[Function]}
    messages={
      Object {
        "Survive The Internet (Party Pack 4)": Object {
          "Max players": 8,
          "Min players": 3,
          "Variants": Array [
            "survive",
            "survive the internet",
            "sti",
          ],
          "chosen": false,
          "locked": false,
          "longName": "Survive The Internet (Party Pack 4)",
          "name": "Survive The Internet",
          "partyPack": "Party Pack 4",
          "time": 1627509347466,
          "username": "dcpesses",
        },
      }
    }
    modList={Array []}
    onDelete={[Function]}
    onMessage={[Function]}
    openQueueHandler={[Function]}
    playerExitHandler={[Function]}
    setNextGame={[Function]}
    startGame={[Function]}
    upcomingGames={Array []}
  />
  <div
    className="left-column"
  >
    <h2>
      Game Requests
    </h2>
    <h4>
      <span
        className="subheading-game"
      >
        Type e.g. 
        <b>
          "!request Blather Round"
        </b>
         in 
        sirgoosewell
        's chat to add
      </span>
    </h4>
    <div
      className="left-column-body"
    >
      <Sidebar
        changeNextGameIdx={[Function]}
        history={Array []}
        moveNextGameBack={[Function]}
        moveNextGameFwd={[Function]}
        nextGameIdx={0}
        requestMode="game"
        togglePlayerSelect={[Function]}
      />
      <div
        className="left-column-inner-body"
      >
        <GameRequest
          gameName="Survive The Internet (Party Pack 4)"
          getActivity={[Function]}
          metadata={
            Object {
              "Max players": 8,
              "Min players": 3,
              "Variants": Array [
                "survive",
                "survive the internet",
                "sti",
              ],
              "chosen": false,
              "locked": false,
              "longName": "Survive The Internet (Party Pack 4)",
              "name": "Survive The Internet",
              "partyPack": "Party Pack 4",
              "time": 1627509347466,
              "username": "dcpesses",
            }
          }
          onDelete={[Function]}
          toggleLock={[Function]}
        />
      </div>
    </div>
  </div>
  <div
    className="right-column"
    width="50px"
  >
    <div
      className="wheel-wrapper"
    >
      <default
        contrastColor="black"
        downDuration={1000}
        fontFamily="Arial"
        isOnlyOnce={false}
        onFinished={[Function]}
        primaryColor="white"
        segColors={
          Array [
            "#99b0fc",
            "#b4a7f2",
            "#beb4f7",
            "#cdc2f9",
            "#aec2f9",
            "#a4d6f9",
          ]
        }
        segments={
          Array [
            "Survive The Internet (Party Pack 4)",
          ]
        }
        size={250}
        upDuration={100}
      />
    </div>
  </div>
  <button
    className="logout"
    onClick={[MockFunction]}
  >
    Logout ➧
  </button>
</div>
`;

exports[`MainScreen should render the player selection screen 1`] = `
<div
  className="player-select"
  id="main-screen"
>
  <MessageHandler
    access_token="foo"
    addGameRequest={[Function]}
    caniplayHandler={[Function]}
    changeNextGameIdx={[Function]}
    channel="sirgoosewell"
    clearQueueHandler={[Function]}
    closeQueueHandler={[Function]}
    messages={
      Object {
        "Survive The Internet (Party Pack 4)": Object {
          "Max players": 8,
          "Min players": 3,
          "Variants": Array [
            "survive",
            "survive the internet",
            "sti",
          ],
          "chosen": false,
          "locked": false,
          "longName": "Survive The Internet (Party Pack 4)",
          "name": "Survive The Internet",
          "partyPack": "Party Pack 4",
          "time": 1627509347466,
          "username": "dcpesses",
        },
      }
    }
    modList={Array []}
    onDelete={[Function]}
    onMessage={[Function]}
    openQueueHandler={[Function]}
    playerExitHandler={[Function]}
    setNextGame={[Function]}
    startGame={[Function]}
    upcomingGames={Array []}
  />
  <div
    className="left-column"
  >
    <h2>
      Seat Requests
    </h2>
    <h4>
      <span
        className="subheading-player"
      >
        Type 
        <b>
          !new
        </b>
         in 
        sirgoosewell
        's chat if you want to join the next game
      </span>
    </h4>
    <div
      className="left-column-body"
    >
      <Sidebar
        changeNextGameIdx={[Function]}
        history={Array []}
        moveNextGameBack={[Function]}
        moveNextGameFwd={[Function]}
        nextGameIdx={0}
        requestMode="seat"
        togglePlayerSelect={[Function]}
      />
      <div
        className="left-column-inner-body"
      >
        <PlayerSelect
          startGame={[Function]}
        />
      </div>
    </div>
  </div>
  <button
    className="logout"
    onClick={[MockFunction]}
  >
    Logout ➧
  </button>
</div>
`;
